README.md
# `pubchem.rs` [![Star me](https://img.shields.io/github/stars/althonos/pubchem.rs.svg?style=social&label=Star&maxAge=3600)](https://github.com/althonos/pubchem.rs/stargazers)

*Rust data structures and client for the [PubChem](https://pubchem.ncbi.nlm.nih.gov/) [REST API](https://pubchemdocs.ncbi.nlm.nih.gov/pug-rest).*

[![Actions](https://img.shields.io/github/workflow/status/althonos/pubchem.rs/Test?style=flat-square&maxAge=600)](https://github.com/althonos/pubchem.rs/actions)
[![Codecov](https://img.shields.io/codecov/c/gh/althonos/pubchem.rs/master.svg?style=flat-square&maxAge=600)](https://codecov.io/gh/althonos/pubchem.rs)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square&maxAge=2678400)](https://choosealicense.com/licenses/mit/)
[![Source](https://img.shields.io/badge/source-GitHub-303030.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/pubchem.rs)
[![Crate](https://img.shields.io/crates/v/pubchem.svg?maxAge=600&style=flat-square)](https://crates.io/crates/pubchem)
[![Documentation](https://img.shields.io/badge/docs.rs-latest-4d76ae.svg?maxAge=2678400&style=flat-square)](https://docs.rs/pubchem)
[![Changelog](https://img.shields.io/badge/keep%20a-changelog-8A0707.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/pubchem.rs/blob/master/CHANGELOG.md)
[![GitHub issues](https://img.shields.io/github/issues/althonos/pubchem.rs.svg?style=flat-square&maxAge=600)](https://github.com/althonos/pubchem.rs/issues)


## 🔌 Usage

### 💊 Compound

Create a `Compound` to query the PubChem API for a single compound. It can
be constructed from a compound ID, from a compound name, from an InChI or
InChIKey, or from a SMILES string:

```rust
extern crate pubchem;

let alanine = pubchem::Compound::new(5950);
let aspirin = pubchem::Compound::with_name("aspirin");
let acetone = pubchem::Compound::with_inchi("InChI=1S/C3H6O/c1-3(2)4/h1-2H3");
let lysine  = pubchem::Compound::with_inchikey("KDXKERNSBIXSRK-YFKPBYRVSA-N");
let benzene = pubchem::Compound::with_smiles("C1=CC=CC=C1");
```

Use the methods to query the REST API with [`ureq`](https://crates.io/crates/ureq).
Dedicated methods exist for common single properties:

```rust
let alanine = pubchem::Compound::new(5950);

alanine.title().unwrap(); // "Alanine"
alanine.molecular_formula().unwrap(); // "C3H7NO2"
alanine.canonical_smiles().unwrap(); // "CC(C(=O)O)N"
alanine.isomeric_smiles().unwrap();  // "C[C@@H](C(=O)O)N"
```

Each method will perform a single query to the PubChem API, which is inefficient
if you wish to retrieve several properties at once. In that case, use the
`properties` method and select which properties you want to retrieve
in a single query:

```rust
use pubchem::CompoundProperty::*;

let properties = pubchem::Compound::new(5950)
    .properties(&[Title, MolecularFormula, CanonicalSMILES])
    .unwrap();

properties.molecular_formula; // Some("C3H7NO2")
properties.canonical_smiles; // Some("CC(C(=O)O)N")
properties.isomeric_smiles; // Some("C[C@@H](C(=O)O)N")
```

To retrieve metadata from multiple compounds at once, use the `Compounds`
struct and use the `properties` method to pack everything into a single
query:

```rust
use pubchem::CompoundProperty::*;

// retrieve metadata from the three aromatic L-amino acids at once
for prop in pubchem::Compounds::new([6140, 145742, 6305])
    .properties(&[Title, IUPACName, ExactMass])
    .unwrap()
{
    println!(
        "[{cid}] {title} {iupac} {mass}g/mol",
        cid = prop.cid,
        title = prop.title.unwrap(),
        iupac = prop.iupac_name.unwrap(),
        mass = prop.exact_mass.unwrap(),
    );
}
```

## 💭 Feedback

### ⚠️ Issue Tracker

Found a bug ? Have an enhancement request ? Head over to the [GitHub issue
tracker](https://github.com/althonos/pubchem.rs/issues) if you need to report
or ask something. If you are filing in on a bug, please include as much
information as you can about the issue, and try to recreate the same bug
in a simple, easily reproducible situation.

<!-- ### 🏗️ Contributing

Contributions are more than welcome! See
[`CONTRIBUTING.md`](https://github.com/althonos/pubchem.rs/blob/main/CONTRIBUTING.md)
for more details. -->

## 📋 Changelog

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
and provides a [changelog](https://github.com/althonos/pubchem.rs/blob/master/CHANGELOG.md)
in the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) format.

## 🔍 See Also

If you're a bioinformatician and a Rustacean, you may be interested in these
other libraries:

- [`uniprot.rs`](https://github.com/althonos/uniprot.rs): Rust data structures
  for the UniProtKB databases.
- [`obofoundry.rs`](https://github.com/althonos/obofoundry.rs): Rust data
  structures for the OBO Foundry.
- [`fastobo`](https://github.com/fastobo/fastobo): Rust parser and abstract
  syntax tree for Open Biomedical Ontologies.

## 📜 License

This library is provided under the open-source
[MIT license](https://choosealicense.com/licenses/mit/).

*This project is in no way not affiliated, sponsored, or otherwise endorsed
by the PubChem developers. It was developed
by [Martin Larralde](https://github.com/althonos/) during his PhD project
at the [European Molecular Biology Laboratory](https://www.embl.de/) in
the [Zeller team](https://github.com/zellerlab).*


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
[Unreleased]: https://github.com/althonos/pubchem.rs/compare/v0.1.1...HEAD

## [v0.1.1] - 2021-01-15
[v0.1.1]: https://github.com/althonos/pubchem.rs/compare/v0.1.0...v0.1.1

### Added
- `pubchem::Compound::synonyms` to retrieve synonym names for a compound.
- `pubchem::Compounds` to query properties from multiple compounds.
- `Deref`, `DerefMut` and `IntoIterator` implementations for `pubchem::model::rest::PropertyTable`.

## [v0.1.0] - 2021-01-15
[v0.1.0]: https://github.com/althonos/pubchem.rs/compare/7f3bb8b...v0.1.0

Initial release.


src/error.rs
//! Ubiquitous types for error management.

use std::io::Error as IoError;
use std::num::ParseFloatError;
use std::num::ParseIntError;

use quick_xml::Error as XmlError;
use thiserror::Error;
use ureq::Error as UreqError;

#[derive(Debug, Error, Clone, PartialEq, Eq)]
/// An error returned by the Power User Gateway REST API.
pub enum ApiError {
    #[error("bad request: {0}")]
    /// Request is improperly formed.
    BadRequest(String),
    #[error("not found: {0}")]
    /// The input record was not found.
    NotFound(String),
    #[error("not allowed: {0}")]
    /// Request not allowed.
    NotAllowed(String),
    #[error("timeout: {0}")]
    /// The request timed out, from server overload or too broad a request.
    Timeout(String),
    #[error("server busy: {0}")]
    /// Too many requests or server is busy, retry later.
    ServerBusy(String),
    #[error("unimplemented!(): {0}")]
    /// The requested operation has not (yet) been implemented by the server.
    Unimplemented(String),
    #[error("server error: {0}")]
    /// Some problem on the server side (such as a database server down).
    ServerError(String),
    #[error("unknown error: {0}")]
    /// An unknown error occurred
    Unknown(String),
}

impl From<crate::model::rest::Fault> for ApiError {
    fn from(fault: crate::model::rest::Fault) -> Self {
        match fault.code.as_str() {
            "PUGREST.BadRequest" => ApiError::BadRequest(fault.message),
            "PUGREST.NotFound" => ApiError::NotFound(fault.message),
            "PUGREST.NotAllowed" => ApiError::NotAllowed(fault.message),
            "PUGREST.Timeout" => ApiError::Timeout(fault.message),
            "PUGREST.ServerBusy" => ApiError::ServerBusy(fault.message),
            "PUGREST.Unimplemented" => ApiError::Unimplemented(fault.message),
            "PUGREST.ServerError" => ApiError::ServerError(fault.message),
            _ => ApiError::Unknown(fault.message),
        }
    }
}

// ---------------------------------------------------------------------------

#[derive(Debug, Error, Clone, PartialEq, Eq)]
/// An error raised by a `FromStr` implementor.
pub enum ParseError {
    #[error(transparent)]
    Int(#[from] ParseIntError),
    #[error(transparent)]
    Float(#[from] ParseFloatError),
}

// ---------------------------------------------------------------------------

#[derive(Debug, Error)]
/// The main error type for the [`pubchem`] crate.
///
/// [`pubchem`]: ../index.html
pub enum Error {
    #[error(transparent)]
    /// The PubChem API responded with an error.
    Api(#[from] ApiError),
    #[error(transparent)]
    /// The HTTP client encountered an error.
    Request(#[from] UreqError),
    #[error(transparent)]
    /// The XML parser encountered an error.
    ///
    /// *Any error from the underlying reader will be wrapped in the
    /// [`XmlError::Io`] variant.*
    ///
    /// [`XmlError::Io`]: https://docs.rs/quick-xml/latest/quick_xml/enum.Error.html#variant.Io
    Xml(#[from] XmlError),
    #[error(transparent)]
    /// A parser returned an error.
    Parse(#[from] ParseError),
}

impl From<IoError> for Error {
    fn from(e: IoError) -> Self {
        Self::from(XmlError::Io(e))
    }
}

impl From<ParseIntError> for Error {
    fn from(e: ParseIntError) -> Self {
        Self::Parse(ParseError::Int(e))
    }
}

impl From<ParseFloatError> for Error {
    fn from(e: ParseFloatError) -> Self {
        Self::Parse(ParseError::Float(e))
    }
}

/// The main result type for the [`pubchem`] crate.
///
/// [`pubchem`]: ../index.html
pub type Result<T> = std::result::Result<T, Error>;


src/utils.rs
use std::fmt::Display;
use std::fmt::Write;
use std::iter::Iterator;

macro_rules! named_enum {
    ($(#[$attributes:meta])*
        pub enum $name:ident {
        $(#[$doc:meta] $variant:ident),*,
    }) => {
        $(#[$attributes])*
        pub enum $name {
            $(#[$doc] $variant),*
        }

        impl $name {
            /// Get the name of the enum variant.
            pub fn name(&self) -> &'static str {
                match self {
                    $($name::$variant => stringify!($variant)),*
                }
            }
        }
    };
}

pub trait Join {
    fn join(&mut self, separator: &str) -> String;
}

impl<I> Join for I
where
    I: Iterator,
    I::Item: Display,
{
    fn join(&mut self, separator: &str) -> String {
        let mut joined = String::new();
        for (i, item) in self.enumerate() {
            if i > 0 {
                joined.push_str(separator);
            }
            write!(joined, "{}", item).unwrap();
        }
        joined
    }
}


src/parser/mod.rs
#[macro_use]
mod macros;

use std::io::BufRead;

use quick_xml::events::BytesStart;
use quick_xml::events::Event;
use quick_xml::Error as XmlError;
use quick_xml::Reader;

use crate::error::Error;

/// A trait for types that can be parsed from an XML element.
pub trait FromXml: Sized {
    fn from_xml<B: BufRead>(
        event: &BytesStart,
        reader: &mut Reader<B>,
        buffer: &mut Vec<u8>,
    ) -> Result<Self, Error>;
}

/// A trait for types that can be returned by the REST API.
pub trait FromApiResponse: FromXml {
    fn from_api_response(response: ureq::Response) -> Result<Self, Error>;
}

impl<T: FromXml> FromApiResponse for T {
    fn from_api_response(response: ureq::Response) -> Result<Self, Error> {
        let reader = response.into_reader();
        let mut xml = Reader::from_reader(std::io::BufReader::new(reader));
        let mut buffer = Vec::new();

        let event = loop {
            buffer.clear();
            match xml.read_event(&mut buffer) {
                Err(e) => return Err(Error::from(e)),
                Ok(Event::Start(e)) => break e.to_owned(),
                Ok(Event::Eof) => {
                    let e = String::from("xml");
                    return Err(Error::from(XmlError::UnexpectedEof(e)));
                }
                _ => (),
            }
        };

        T::from_xml(&event, &mut xml, &mut buffer)
    }
}


src/parser/macros.rs
#[allow(unused_macros)]
macro_rules! parse_inner {
    ($event:expr, $reader:expr, $buffer:expr, $($rest:tt)*) => ({
        #[allow(unused_imports)]
        loop {
            use $crate::quick_xml::events::BytesEnd;
            use $crate::quick_xml::events::BytesStart;
            use $crate::quick_xml::events::Event;
            use $crate::quick_xml::Error as XmlError;

            $buffer.clear();
            match $reader.read_event($buffer) {
                Ok(Event::Start(ref x)) => {
                    parse_inner_impl!(x, x.local_name(), $($rest)*);
                    $reader.read_to_end(x.local_name(), &mut Vec::new())?;
                    unimplemented!(
                        "`{}` in `{}`",
                        String::from_utf8_lossy(x.local_name()),
                        String::from_utf8_lossy($event.local_name())
                    );
                }
                Err(e) => {
                    return Err(Error::from(e));
                }
                Ok(Event::Eof) => {
                    let e = String::from_utf8_lossy($event.local_name()).to_string();
                    return Err(Error::from(XmlError::UnexpectedEof(e)));
                }
                Ok(Event::End(ref e)) if e.local_name() == $event.local_name() => {
                    break;
                }
                Ok(Event::End(ref e)) => {
                    let expected = $event.unescaped()
                        .map(|s| String::from_utf8_lossy(s.as_ref()).to_string())?;
                    let found = String::from_utf8_lossy(e.name()).to_string();
                    let e = XmlError::EndEventMismatch { expected, found };
                    return Err(Error::from(e));
                }
                _ => continue,
            }
        }
    })
}

#[allow(unused_macros)]
macro_rules! parse_inner_impl {
    ( $x:ident, $name:expr ) => ();
    ( $x:ident, $name:expr, ) => ();
    ( $x:ident, $name:expr, $e:ident @ $l:expr => $r:expr ) => (
        if $name == $l {
            let $e = $x.clone().into_owned();
            $r;
            continue;
        }
    );
    ( $x:ident, $name:expr, $l:expr => $r:expr ) => (
        if $name == $l {
            $r;
            continue;
        }
    );
    ( $x:ident, $name:expr, $e:ident @ $l:expr => $r:expr, $($rest:tt)*) => (
        parse_inner_impl!( $x, $name, $e @ $l => $r );
        parse_inner_impl!( $x, $name, $($rest)* );
    );
    ( $x:ident, $name:expr, $l:expr => $r:expr, $($rest:tt)*) => (
        parse_inner_impl!( $x, $name, $l => $r );
        parse_inner_impl!( $x, $name, $($rest)* );
    )
}

#[allow(unused_macros)]
macro_rules! parse_comment {
    ( $event:ident, $reader:ident, $buffer:ident, $comment:ident ) => {
        parse_comment!{$event, $reader, $buffer, $comment, }
    };
    ( $event:ident, $reader:ident, $buffer:ident, $comment:ident, $($rest:tt)* ) => {
        parse_inner!{$event, $reader, $buffer,
            b"text" => {
                $comment.text.push($reader.read_text(b"text", $buffer)?);
            },
            m @ b"molecule" => {
                $comment.molecule = Molecule::from_xml(&m, $reader, $buffer)
                    .map(Some)?;
            },
            $($rest)*
        }
    }
}


src/lib.rs
#![doc = include_str!("../README.md")]

extern crate form_urlencoded;
extern crate quick_xml;
extern crate thiserror;
extern crate ureq;

#[macro_use]
mod parser;
#[macro_use]
mod utils;

pub mod error;
pub mod model;

use std::borrow::Cow;

use self::error::Error;
use self::model::rest;
use self::parser::FromApiResponse;
use self::utils::Join;

named_enum! {
    #[derive(Debug, PartialEq, Eq)]
    /// A single property that can be retrieved from a compound.
    pub enum CompoundProperty {
        /// Molecular formula.
        MolecularFormula,
        /// The molecular weight is the sum of all atomic weights of the constituent atoms in a compound, measured in g/mol. In the absence of explicit isotope labelling, averaged natural abundance is assumed. If an atom bears an explicit isotope label, 100% isotopic purity is assumed at this location.
        MolecularWeight,
        /// Canonical SMILES (Simplified Molecular Input Line Entry System) string.  It is a unique SMILES string of a compound, generated by a “canonicalization” algorithm.
        CanonicalSMILES,
        /// Isomeric SMILES string.  It is a SMILES string with stereochemical and isotopic specifications.
        IsomericSMILES,
        /// Standard IUPAC International Chemical Identifier (InChI).  It does not allow for user selectable options in dealing with the stereochemistry and tautomer layers of the InChI string.
        InChI,
        /// Hashed version of the full standard InChI, consisting of 27 characters.
        InChIKey,
        /// Chemical name systematically determined according to the IUPAC nomenclatures.
        IUPACName,
        /// The title used for the compound summary page.
        Title,
        /// Computationally generated octanol-water partition coefficient or distribution coefficient. XLogP is used as a measure of hydrophilicity or hydrophobicity of a molecule.
        XLogP,
        /// The mass of the most likely isotopic composition for a single molecule, corresponding to the most intense ion/molecule peak in a mass spectrum.
        ExactMass,
        /// The mass of a molecule, calculated using the mass of the most abundant isotope of each element.
        MonoisotopicMass,
        /// Topological polar surface area, computed by the algorithm described in the paper by Ertl et al.
        TPSA,
        /// The molecular complexity rating of a compound, computed using the Bertz/Hendrickson/Ihlenfeldt formula.
        Complexity,
        /// The total (or net) charge of a molecule.
        Charge,
        /// Number of hydrogen-bond donors in the structure.
        HBondDonorCount,
        /// Number of hydrogen-bond acceptors in the structure.
        HBondAcceptorCount,
        /// Number of rotatable bonds.
        RotatableBondCount,
        /// Number of non-hydrogen atoms.
        HeavyAtomCount,
        /// Number of atoms with enriched isotope(s)
        IsotopeAtomCount,
        /// Total number of atoms with tetrahedral (sp3) stereo [e.g., (R)- or (S)-configuration]
        AtomStereoCount,
        /// Number of atoms with defined tetrahedral (sp3) stereo.
        DefinedAtomStereoCount,
        /// Number of atoms with undefined tetrahedral (sp3) stereo.
        UndefinedAtomStereoCount,
        /// Total number of bonds with planar (sp2) stereo [e.g., (E)- or (Z)-configuration].
        BondStereoCount,
        /// Number of atoms with defined planar (sp2) stereo.
        DefinedBondStereoCount,
        /// Number of atoms with undefined planar (sp2) stereo.
        UndefinedBondStereoCount,
        /// Number of covalently bound units.
        CovalentUnitCount,
        /// Analytic volume of the first diverse conformer (default conformer) for a compound.
        Volume3D,
        /// The x component of the quadrupole moment (Qx) of the first diverse conformer (default conformer) for a compound.
        XStericQuadrupole3D,
        /// The y component of the quadrupole moment (Qy) of the first diverse conformer (default conformer) for a compound.
        YStericQuadrupole3D,
        /// The z component of the quadrupole moment (Qz) of the first diverse conformer (default conformer) for a compound.
        ZStericQuadrupole3D,
        /// Total number of 3D features (the sum of FeatureAcceptorCount3D, FeatureDonorCount3D, FeatureAnionCount3D, FeatureCationCount3D, FeatureRingCount3D and FeatureHydrophobeCount3D)
        FeatureCount3D,
        /// Number of hydrogen-bond acceptors of a conformer.
        FeatureAcceptorCount3D,
        /// Number of hydrogen-bond donors of a conformer.
        FeatureDonorCount3D,
        /// Number of anionic centers (at pH 7) of a conformer.
        FeatureAnionCount3D,
        /// Number of cationic centers (at pH 7) of a conformer.
        FeatureCationCount3D,
        /// Number of rings of a conformer.
        FeatureRingCount3D,
        /// Number of hydrophobes of a conformer.
        FeatureHydrophobeCount3D,
        /// Conformer sampling RMSD in Å.
        ConformerModelRMSD3D,
        /// Total number of 3D features (the sum of FeatureAcceptorCount3D, FeatureDonorCount3D, FeatureAnionCount3D, FeatureCationCount3D, FeatureRingCount3D and FeatureHydrophobeCount3D)
        EffectiveRotorCount3D,
        /// The number of conformers in the conformer model for a compound.
        ConformerCount3D,
        /// Base64-encoded PubChem Substructure Fingerprint of a molecule.
        Fingerprint2D,
    }
}

#[derive(Debug)]
/// A client for retrieving information about a single PubChem compound.
pub struct Compound {
    namespace: Cow<'static, str>,
    identifier: Cow<'static, str>,
}

impl Compound {
    /// Create a new `Compound` from the given compound ID.
    pub fn new(id: u32) -> Self {
        Self {
            namespace: Cow::Borrowed("cid"),
            identifier: Cow::Owned(id.to_string()),
        }
    }

    /// Create a new `Compound` matching the given compound name.
    pub fn with_name(name: &str) -> Self {
        Self {
            namespace: Cow::Borrowed("name"),
            identifier: Cow::Owned(name.to_string()),
        }
    }

    /// Create a new `Compound` matching the given SMILES.
    pub fn with_smiles(smiles: &str) -> Self {
        Self {
            namespace: Cow::Borrowed("smiles"),
            identifier: Cow::Owned(smiles.to_string()),
        }
    }

    /// Create a new `Compound` matching the given InChI.
    pub fn with_inchi(inchi: &str) -> Self {
        Self {
            namespace: Cow::Borrowed("inchi"),
            identifier: Cow::Owned(inchi.to_string()),
        }
    }

    /// Create a new `Compound` matching the given InChIKey.
    pub fn with_inchikey(inchikey: &str) -> Self {
        Self {
            namespace: Cow::Borrowed("inchikey"),
            identifier: Cow::Owned(inchikey.to_string()),
        }
    }

    // pub fn with_sdf(sdf: &str) -> Self {
    //     Self {
    //         namespace: Cow::Borrowed("sdf"),
    //         identifiers: Cow::Owned(sdf.to_string())
    //     }
    // }

    /// Request the REST API for the given operation.
    ///
    /// The response is checked to see if the HTTP client or the API errored,
    /// otherwise the raw response is returned so that it can be parsed by
    /// the appropriate method.
    ///
    fn request(&self, operation: &str) -> Result<ureq::Response, Error> {
        let url = format!(
            "https://pubchem.ncbi.nlm.nih.gov/rest/pug/{dom}/{ns}/{op}/XML",
            dom = "compound",
            ns = &self.namespace,
            op = operation
        );
        let form_data = form_urlencoded::Serializer::new(String::new())
            .append_pair(&self.namespace, &self.identifier)
            .finish();
        match ureq::post(&url)
            .set("Accept", "application/xml")
            .set("Content-Type", "application/x-www-form-urlencoded")
            .send_string(&form_data)
        {
            Err(ureq::Error::Status(400 | 404 | 405 | 500 | 501 | 503 | 504, response)) => {
                let fault = rest::Fault::from_api_response(response)?;
                Err(Error::Api(fault.into()))
            }
            Err(e) => Err(Error::Request(e)),
            Ok(response) => Ok(response),
        }
    }

    /// Retrieve several properties at once for the compound.
    pub fn properties<'p, P>(&self, properties: P) -> Result<rest::Properties, Error>
    where
        P: IntoIterator<Item = &'p CompoundProperty>,
    {
        let mut path = String::from("/property/");
        path.push_str(&properties.into_iter().map(CompoundProperty::name).join(","));
        self.request(&path)
            .and_then(|response| rest::PropertyTable::from_api_response(response))
            .map(|mut table| table.properties.pop().unwrap())
    }

    /// Retrieve the main PubChem designation for the compound.
    pub fn title(&self) -> Result<String, Error> {
        let properties = self.properties(&[CompoundProperty::Title])?;
        Ok(properties
            .title
            .expect("All PubChem compounds should have a title."))
    }

    /// Retrieve the molecular formula of the compound.
    ///
    /// # Example
    /// ```
    /// let compound = pubchem::Compound::with_name("aspirin");
    /// assert_eq!(compound.molecular_formula().unwrap(), "C9H8O4");
    /// ```
    pub fn molecular_formula(&self) -> Result<String, Error> {
        let properties = self.properties(&[CompoundProperty::MolecularFormula])?;
        Ok(properties
            .molecular_formula
            .expect("All PubChem compounds should have a formula."))
    }

    /// Retrieve the canonical SMILES string for the compound.
    pub fn canonical_smiles(&self) -> Result<String, Error> {
        let properties = self.properties(&[CompoundProperty::CanonicalSMILES])?;
        Ok(properties
            .canonical_smiles
            .expect("All PubChem compounds should have a SMILES."))
    }

    /// Retrieve the isomeric SMILES string for the compound.
    ///
    /// # Example
    /// ```
    /// let alanine = pubchem::Compound::with_name("alanine");
    /// assert_eq!(alanine.isomeric_smiles().unwrap(), "C[C@H](C(=O)O)N");
    /// ```
    pub fn isomeric_smiles(&self) -> Result<String, Error> {
        let properties = self.properties(&[CompoundProperty::IsomericSMILES])?;
        Ok(properties
            .isomeric_smiles
            .expect("All PubChem compounds should have a SMILES."))
    }

    // /// Retrieve the entire PubChem record for the compound.
    // pub fn record(&self) {
    //     unimplemented!()
    // }

    /// Retrieve synonym names for the compound.
    pub fn synonyms(&self) -> Result<Vec<String>, Error> {
        self.request("synonyms")
            .and_then(|response| rest::InformationList::from_api_response(response))
            .map(|mut list| list.informations.pop().unwrap().synonyms)
    }

    /// Retrieve the Compound IDs designating the compound.
    pub fn cids(&self) -> Result<Vec<i32>, Error> {
        self.request("cids")
            .and_then(|response| rest::IdentifierList::from_api_response(response))
            .map(|list| list.cids)
    }

    /// Retrieve the Substance IDs associated with the compound.
    pub fn sids(&self) -> Result<Vec<i32>, Error> {
        self.request("sids")
            .and_then(|response| rest::InformationList::from_api_response(response))
            .map(|mut list| list.informations.pop().unwrap().sids)
    }

    /// Retrieve the Assay IDs associated with the compound.
    pub fn aids(&self) -> Result<Vec<i32>, Error> {
        self.request("aids")
            .and_then(|response| rest::InformationList::from_api_response(response))
            .map(|mut list| list.informations.pop().unwrap().aids)
    }

    // pub fn assay_summary(&self) {
    //
    // }
    //
    // pub fn classification(&self) {
    //
    // }
    //
    // pub fn xref(&self) {
    //
    // }
    //
    // pub fn description(&self) {
    //
    // }
    //
    // pub fn conformers(&self) {
    //
    // }
}

#[derive(Debug)]
/// A client for retrieving information about multiple PubChem compoumds at once.
pub struct Compounds {
    namespace: Cow<'static, str>,
    identifiers: Cow<'static, str>,
}

impl Compounds {
    /// Query several compounds with the given compound IDs.
    pub fn new<I: IntoIterator<Item = u32>>(ids: I) -> Self {
        Self {
            namespace: Cow::Borrowed("cid"),
            identifiers: Cow::Owned(ids.into_iter().join(",")),
        }
    }

    /// Request the REST API for the given operation.
    ///
    /// The response is checked to see if the HTTP client or the API errored,
    /// otherwise the raw response is returned so that it can be parsed by
    /// the appropriate method.
    ///
    fn request(&self, operation: &str) -> Result<ureq::Response, Error> {
        let url = format!(
            "https://pubchem.ncbi.nlm.nih.gov/rest/pug/{dom}/{ns}/{op}/XML",
            dom = "compound",
            ns = &self.namespace,
            op = operation
        );
        let form_data = form_urlencoded::Serializer::new(String::new())
            .append_pair(&self.namespace, &self.identifiers)
            .finish();
        match ureq::post(&url)
            .set("Accept", "application/xml")
            .set("Content-Type", "application/x-www-form-urlencoded")
            .send_string(&form_data)
        {
            Err(ureq::Error::Status(400 | 404 | 405 | 500 | 501 | 503 | 504, response)) => {
                let fault = rest::Fault::from_api_response(response)?;
                Err(Error::Api(fault.into()))
            }
            Err(e) => Err(Error::Request(e)),
            Ok(response) => Ok(response),
        }
    }

    /// Retrieve several properties at once for the compounds.
    pub fn properties<'p, P>(&self, properties: P) -> Result<rest::PropertyTable, Error>
    where
        P: IntoIterator<Item = &'p CompoundProperty>,
    {
        let mut path = String::from("/property/");
        path.push_str(&properties.into_iter().map(CompoundProperty::name).join(","));
        self.request(&path)
            .and_then(|response| rest::PropertyTable::from_api_response(response))
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    use crate::error::ApiError;

    #[test]
    fn compound_new() {
        let compound = Compound::new(2244);
        let properties = compound
            .properties(&[CompoundProperty::Title])
            .expect("compound property retrieval should not fail");
        assert_eq!(properties.title.as_ref().unwrap(), "Aspirin");
    }

    #[test]
    fn compound_with_name() {
        let compound = Compound::with_name("lyciumin A");
        let properties = compound
            .properties(&[CompoundProperty::Title])
            .expect("compound property retrieval should not fail");
        assert_eq!(properties.cid, 14430290);
    }

    #[test]
    fn compound_with_smiles() {
        let compound = Compound::with_smiles("CC(=O)OC1=CC=CC=C1C(=O)O");
        let properties = compound
            .properties(&[CompoundProperty::Title])
            .expect("compound property retrieval should not fail");
        assert_eq!(properties.cid, 2244);
    }

    #[test]
    fn compound_with_inchikey() {
        let compound = Compound::with_inchikey("AUJXLBOHYWTPFV-UHFFFAOYSA-N");
        assert_eq!(compound.title().unwrap(), "Echinomycin");
    }

    #[test]
    fn compound_with_inchi() {
        let compound = Compound::with_inchi("InChI=1S/C3H6O/c1-3(2)4/h1-2H3");
        assert_eq!(compound.title().unwrap(), "Acetone");
    }

    #[test]
    fn compound_cids() {
        let compound = Compound::new(2244);
        assert_eq!(compound.cids().unwrap(), vec![2244])
    }

    #[test]
    #[rustfmt::skip]
    fn compound_sids() {
        let compound = Compound::new(10444160);
        assert_eq!(
            compound.sids().unwrap(),
            vec![
                 15464793,  40571804,  50320706, 103261147,
                319343201, 383830942, 386266192, 459034771
            ]
        )
    }

    #[test]
    fn compound_synonyms() {
        let compound = Compound::new(180);
        let synonyms = compound.synonyms().unwrap();
        assert_eq!(synonyms.len(), 578);
        assert_eq!(&synonyms[0], "acetone");
        assert_eq!(&synonyms[1], "2-propanone");
    }

    #[test]
    fn compound_name_not_found() {
        let compound = Compound::with_name("none");
        match compound.cids() {
            Err(Error::Api(ApiError::NotFound(_))) => (),
            Err(e) => panic!("unexpected error {}", e),
            Ok(_) => panic!("unexpected success"),
        }
    }

    #[test]
    fn compounds_properties() {
        let compounds = Compounds::new([6140, 6057, 6305]);
        let property_table = compounds.properties(&[CompoundProperty::Title]).unwrap();
        assert_eq!(property_table.properties.len(), 3);
        for properties in property_table.properties {
            match properties.cid {
                6140 => assert_eq!(properties.title.unwrap(), "Phenylalanine"),
                6057 => assert_eq!(properties.title.unwrap(), "Tyrosine"),
                6305 => assert_eq!(properties.title.unwrap(), "Tryptophan"),
                _ => unreachable!(),
            }
        }
    }
}
